ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveAntiSemiJoinRule.java

```diff
@@ -135,43 +142,76 @@ private List<RexNode> getResidualFilterNodes(Filter filter, Join join) {
-      if (filterNode.getKind() == SqlKind.IS_NULL) {
-        // Null filter from right side table can be removed and its a pre-condition for anti join conversion.
-        if (HiveCalciteUtil.hasAllExpressionsFromRightSide(join, Collections.singletonList(filterNode))
-            && isStrong(((RexCall) filterNode).getOperands().get(0))) {
-          hasNullFilterOnRightSide = true;
-        } else {
-          filterList.add(filterNode);
-        }
-      } else {
-        if (HiveCalciteUtil.hasAnyExpressionFromRightSide(join, Collections.singletonList(filterNode))) {
-          // If some non null condition is present from right side, we can not convert the join to anti join as
-          // anti join does not project the fields from right side.
-          return null;
-        } else {
-          filterList.add(filterNode);
-        }
+      final ImmutableBitSet usedFields = RelOptUtil.InputFinder.bits(filterNode);
+      boolean usesFieldFromRHS = usedFields.intersects(rhsFields);
+
+      if(!usesFieldFromRHS) {
+        // Only LHS fields or constants, so the filterNode is part of the residual filter
+        filterList.add(filterNode);
+        continue;
+      }
+
+      // In the following we check for filter nodes that let us deduce that
+      // "an (originally) not-null column of RHS IS NULL because the LHS row will not be matched"
+
+      boolean usesRHSFieldsOnly = rhsFields.contains(usedFields);
+      if (!usesRHSFieldsOnly) {
+        // If there is a mix between LHS and RHS fields, don't convert to anti-join
+        return Optional.empty();
+      }
+
+      if(filterNode.getKind() != SqlKind.IS_NULL) {
+        return Optional.empty();
+      }
+
+      // Null filter from right side table can be removed and it is a pre-condition for anti join conversion.
+      RexNode arg = ((RexCall) filterNode).getOperands().get(0);
+      if (isStrong(arg, notNullColumnsFromRightSide)) {
+        hasNullFilterOnRightSide = true;
+      } else if(!isStrong(arg, rhsFields)) {
+        // if all RHS fields are null and the IS NULL is still not fulfilled, bail out
+        return Optional.empty();
       }
     }
 
     if (!hasNullFilterOnRightSide) {
-      return null;
+      return Optional.empty();
+    }
+    return Optional.of(filterList);
+  }
+
+  private ImmutableBitSet getNotNullColumnsFromRightSide(RelNode joinRel) {
+    // we need to shift the indices of the second child to the right
+    int shift = (joinRel.getInput(0)).getRowType().getFieldCount();
+
+    ImmutableBitSet.Builder builder = ImmutableBitSet.builder();
+    List<RelDataTypeField> rhsFields = joinRel.getInput(1).getRowType().getFieldList();
+    for(RelDataTypeField field : rhsFields) {
+      if(!field.getType().isNullable()) {
+        builder.set(shift+field.getIndex());
+      }
     }
-    return filterList;
+    return builder.build();
   }
 
-  private boolean isStrong(RexNode rexNode) {
-    AtomicBoolean hasCast = new AtomicBoolean(false);
```

Comment:
Previously, since AtomicBoolean was used, I am wondering there might have been a good reason.  Was there a concurrency issue that motivated the original implementation.  Changing it to a MutableBoolean means that this variable is not thread safe.

Reply:
A RexVisitor is single-threaded and the `hasCast` is a local variable, so no risk of concurrency issues. See `org.apache.calcite.rex.RexChecker#failCount` for a similar non-synchronized usage in Calcite.

I've stumbled upon `org.apache.calcite.util.Util.FoundOne`, see for example `org.apache.calcite.rex.RexUtil#isDeterministic`, which should be thread safe. I've replaced the hasCast variable with the FoundOne to make it future proof.

Reply:
Yes, I couldn't think of a compelling reason why AtomicBoolean was used in the first place. The updated change looks good.